gitreceive/server.go

```diff
@@ -0,0 +1,373 @@
+/*
+gitreceive handles 'smart' Git HTTP requests for Flynn
+
+This HTTP server can service 'git clone', 'git push' etc. commands
+from Git clients that use the 'smart' Git HTTP protocol (git-upload-pack
+and git-receive-pack).
+
+Derived from https://gitlab.com/gitlab-org/gitlab-git-http-server
+*/
+package main
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"crypto/hmac"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/flynn/flynn/controller/client"
+	"github.com/flynn/flynn/controller/utils"
+	"github.com/flynn/flynn/pkg/archiver"
+	"github.com/flynn/flynn/pkg/status"
+)
+
+func main() {
+	key := os.Getenv("CONTROLLER_KEY")
+	if key == "" {
+		log.Fatal("missing CONTROLLER_KEY env var")
+	}
+	cc, err := controller.NewClient("", key)
+	if err != nil {
+		log.Fatalln("Unable to connect to controller:", err)
+	}
+	log.Fatal(http.ListenAndServe(":"+os.Getenv("PORT"), newGitHandler(cc, []byte(key))))
+}
+
+type gitHandler struct {
+	controller *controller.Client
+	authKey    []byte
+}
+
+type gitService struct {
+	method     string
+	suffix     string
+	handleFunc func(gitEnv, string, string, http.ResponseWriter, *http.Request)
+	rpc        string
+}
+
+type gitEnv struct {
+	App string
+}
+
+// Routing table
+var gitServices = [...]gitService{
+	{"GET", "/info/refs", handleGetInfoRefs, ""},
+	{"POST", "/git-upload-pack", handlePostRPC, "git-upload-pack"},
+	{"POST", "/git-receive-pack", handlePostRPC, "git-receive-pack"},
+}
+
+func newGitHandler(controller *controller.Client, authKey []byte) *gitHandler {
+	return &gitHandler{controller, authKey}
+}
+
+func (h *gitHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var g gitService
+
+	if r.URL.Path == status.Path {
+		status.HealthyHandler.ServeHTTP(w, r)
+		return
+	}
+
+	log.Printf("%s %q", r.Method, r.URL)
```

Comment:
Should we wrap the handler in `httphelper.NewRequestLogger` to standardise our logging? Or at least use log15?